File: 400-lambda.rb
ZERO = ->(f) { ->(v) {       v   } }
ONE  = ->(f) { ->(v) {    f.(v)  } }
TWO  = ->(f) { ->(v) { f.(f.(v)) } }

MULT = ->(a) { ->(b) { ->(f) {  a.(b.(f)) } } }

FOUR = MULT.(TWO).(TWO)
EIGHT = MULT.(FOUR).(TWO)

def to_integer(num)
  num.(->(x) { x + 1}).(0)
end

puts to_integer(FOUR)
puts to_integer(EIGHT)

# Next up - boolean logic

File: js/400-lambda.js
const ZERO = (f) => { return (v) => { return     v   }}
const ONE  = (f) => { return (v) => { return   f(v)  }}
const TWO  = (f) => { return (v) => { return f(f(v)) }}

const MULT = (a) => { return (b) => { return (f) => { return a(b(f)) } } }

const FOUR = MULT(TWO)(TWO)
const EIGHT = MULT(FOUR)(TWO)

function toInt(x) {
  return x((v) => { return v + 1 })(0)
}

console.log(toInt(FOUR))
console.log(toInt(EIGHT))

// Next up - boolean logic
